pkg/utils/utils.go

```diff
@@ -200,8 +201,71 @@ func DiscoverSriovDevices(withUnsupported bool, storeManager StoreManagerInterfa
 }
 
 // SyncNodeState Attempt to update the node state to match the desired state
-func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool) error {
-	return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+func SyncNodeState(newState *sriovnetworkv1.SriovNetworkNodeState, pfsToConfig map[string]bool, parallelNicConfig bool) error {
+	if !parallelNicConfig {
+		return ConfigSriovInterfaces(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+	}
+	return ConfigSriovInterfacesInParallel(newState.Spec.Interfaces, newState.Status.Interfaces, pfsToConfig)
+}
+
+func ConfigSriovInterfacesInParallel(interfaces []sriovnetworkv1.Interface, ifaceStatuses []sriovnetworkv1.InterfaceExt, pfsToConfig map[string]bool) error {
+	log.Log.V(2).Info("ConfigSriovInterfacesInParallel(): start sriov configuration")
+	if IsKernelLockdownMode(true) && hasMellanoxInterfacesInSpec(ifaceStatuses, interfaces) {
+		log.Log.Error(nil, "cannot use mellanox devices when in kernel lockdown mode")
+		return fmt.Errorf("cannot use mellanox devices when in kernel lockdown mode")
+	}
+	// TODO(e0ne): store all errors in SriovNetworkNodeState
+	var result error
+	wg := sync.WaitGroup{}
+	for _, ifaceStatus := range ifaceStatuses {
+		configured := false
+		for _, iface := range interfaces {
+			if iface.PciAddress == ifaceStatus.PciAddress {
+				configured = true
+
+				if skip := pfsToConfig[iface.PciAddress]; skip {
+					break
+				}
+
+				if !NeedUpdate(&iface, &ifaceStatus) {
+					log.Log.V(2).Info("syncNodeState(): no need update interface", "address", iface.PciAddress)
+					break
+				}
+
+				wg.Add(1)
+				go func(iface *sriovnetworkv1.Interface, ifaceStatus *sriovnetworkv1.InterfaceExt) {
+					if err := configSriovDevice(iface, ifaceStatus); err != nil {
+						log.Log.Error(err, "ConfigSriovInterfacesInParallel(): fail to configure sriov interface. resetting interface.", "address", iface.PciAddress)
+						result = err
+						if resetErr := resetSriovDevice(*ifaceStatus); resetErr != nil {
+							log.Log.Error(resetErr, "SyncNodeState(): fail to reset on error SR-IOV interface")
+							result = resetErr
+						}
+					}
+					wg.Done()
+				}(&iface, &ifaceStatus)
+
+				break
+			}
+		}
+		if !configured && ifaceStatus.NumVfs > 0 {
+			if skip := pfsToConfig[ifaceStatus.PciAddress]; skip {
+				continue
+			}
+
+			if err := resetSriovDevice(ifaceStatus); err != nil {
```

Comment:
this can also happen in parallel


i would do an initial calculation of which interfaces needs configuring and which need reset then run both parallel
wait for them to finish and report errors if any.

Reply:
Is it OK to implement it in a follow up PR? if not, I'll push it in an additional commit

Reply:
i feel like reset is also part of NIC configuration. do you see issue to handle it in the same PR ?